hyper-capi/examples/client.c

```diff
@@ -0,0 +1,303 @@
+	printf("%.*s: %.*s\n", (int) name_len, name, (int) value_len, value);
+	return HYPER_ITER_CONTINUE;
+}
+
+static hyper_iter_step print_each_chunk(void *userdata, const hyper_buf *chunk) {
+	const uint8_t *buf = hyper_buf_bytes(chunk);
+	size_t len = hyper_buf_len(chunk);
+
+	write(1, buf, len);
+
+	return HYPER_ITER_CONTINUE;
+}
+
+typedef enum {
+	EXAMPLE_NOT_SET = 0, // tasks we don't know about won't have a userdata set
+	EXAMPLE_HANDSHAKE,
+	EXAMPLE_SEND,
+	EXAMPLE_RESP_BODY
+} example_id;
+
+#define STR_ARG(XX) (uint8_t *)XX, strlen(XX)
+
+int main(int argc, char *argv[]) {
+        const char *host = argc > 1 ? argv[1] : "httpbin.org";
+        const char *port = argc > 2 ? argv[2] : "80";
+        const char *path = argc > 3 ? argv[3] : "/";
+        printf("connecting to port %s on %s...\n", port, host);
+
+        int fd = connect_to(host, port);
+	if (fd < 0) {
+		return 1;
+	}
+        printf("connected to %s, now get %s\n", host, path);
+
+	if (fcntl(fd, F_SETFL, O_NONBLOCK) != 0) {
+		printf("failed to set socket to non-blocking\n");
+		return 1;
+	}
+
+	fd_set fds_read;
+	fd_set fds_write;
+	fd_set fds_excep;
+
+	struct conn_data *conn = malloc(sizeof(struct conn_data));
+
+	conn->fd = fd;
+	conn->read_waker = NULL;
+	conn->write_waker = NULL;
+
+
+	// Hookup the IO
+	hyper_io *io = hyper_io_new();
+	hyper_io_set_data(io, (void *)conn);
+	hyper_io_set_read(io, read_cb);
+	hyper_io_set_write(io, write_cb);
+
+	printf("http handshake ...\n");
+
+	// We need an executor generally to poll futures
+	hyper_executor *exec = hyper_executor_new();
+
+	// Prepare client options
+	hyper_clientconn_options *opts = hyper_clientconn_options_new();
+	hyper_clientconn_options_exec(opts, exec);
+
+	hyper_task *handshake = hyper_clientconn_handshake(io, opts);
+	hyper_task_set_data(handshake, (void *)EXAMPLE_HANDSHAKE);
+
+	// Let's wait for the handshake to finish...
+	hyper_executor_push(exec, handshake);
+
+	// The polling state machine!
+	while (1) {
+		// Poll all ready tasks and act on them...
+		while (1) {
+			hyper_task *task = hyper_executor_poll(exec);
+			if (!task) {
+				break;
+			}
+			switch ((example_id) hyper_task_userdata(task)) {
+			case EXAMPLE_HANDSHAKE:
+				;
+				if (hyper_task_type(task) == HYPER_TASK_ERROR) {
+					printf("handshake error!\n");
+					return 1;
+				}
+				assert(hyper_task_type(task) == HYPER_TASK_CLIENTCONN);
+
+				printf("preparing http request ...\n");
+
+				hyper_clientconn *client = hyper_task_value(task);
+				hyper_task_free(task);
+
+				// Prepare the request
+				hyper_request *req = hyper_request_new();
+				if (hyper_request_set_method(req, STR_ARG("GET"))) {
+					printf("error setting method\n");
+					return 1;
+				}
+				if (hyper_request_set_uri(req, STR_ARG(path))) {
+					printf("error setting uri\n");
+					return 1;
+				}
+
+				hyper_headers *req_headers = hyper_request_headers(req);
+				hyper_headers_set(req_headers,  STR_ARG("host"), STR_ARG(host));
```

Comment:
Question on this: I can use `hyper_headers_set()` if I only add a single header with the name, but if I want to add two headers with the same name I would need to make the second one with `hyper_headers_add` ?

Reply:
Yes, `_set` overwrites any previous value. `_add` appends, even if it's empty. You could think of `_set` as combining `_remove` and `_add` (perhaps `_remove` is missing from the C API...).

So it'd also be fair to have just used `hyper_headers_add` in this case.

Reply:
Ah yes, so I can just use `*_add()`, thanks!

Reply:
... except that `_add()` doesn't seem to be provided yet:
~~~
client.c:(.text+0x666): undefined reference to `hyper_headers_add'
/usr/bin/ld: client.c:(.text+0x68e): undefined reference to `hyper_headers_add'
collect2: error: ld returned 1 exit status